File: src/test/java/com/lucas/matches/service/util/LicenseParserTest.java
package com.lucas.matches.service.util;

import com.lucas.matches.service.exception.InvalidLicenseException;
import com.lucas.matches.service.domain.License;
import com.lucas.matches.service.domain.LicenseType;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class LicenseParserTest {
    private LicenseParser licenseParser;
    @BeforeAll
    public void setup(){
        licenseParser = new LicenseParser();
    }


    @Test
    @DisplayName("Parse a valid tournament license")
    public void parseValidTournamentLicense(){
        String licenseString = "TOURNAMENT-t1";
        List<License> licenses = licenseParser.parse(List.of(licenseString));

        assertNotNull(licenses);
        assertNotNull(licenses.get(0));
        assertEquals(LicenseType.TOURNAMENT, licenses.get(0).getLicenseType());
    }

    @Test
    @DisplayName("Parse a valid match license")
    public void parseValidMatchLicense(){
        String licenseString = "MATCH-m1";
        List<License> licenses = licenseParser.parse(List.of(licenseString));

        assertNotNull(licenses);
        assertNotNull(licenses.get(0));
        assertEquals(LicenseType.MATCH, licenses.get(0).getLicenseType());
    }
    @Test
    @DisplayName("Parse multiple valid licenses")
    public void parseMultipleValidLicenses(){
        List<String> licenseStringList = List.of("TOURNAMENT-t1", "MATCH-m1");
        List<License> licenses = licenseParser.parse(licenseStringList);

        assertNotNull(licenses);
        assertEquals(2, licenses.size());
        assertEquals(LicenseType.TOURNAMENT, licenses.get(0).getLicenseType());
        assertEquals(LicenseType.MATCH, licenses.get(1).getLicenseType());
    }

    @Test
    @DisplayName("Parse an invalid license")
    public void parseInvalidLicense(){
        String licenseString = "invalid-m1";
        assertThrows(InvalidLicenseException.class, ()-> licenseParser.parse(List.of(licenseString)));

    }

    @Test
    @DisplayName("Parse badly formatted license")
    public void parseBadlyFormattedLicense(){
        String licenseString = "MATCHm1";
        assertThrows(InvalidLicenseException.class, ()-> licenseParser.parse(List.of(licenseString)));

    }


}

File: src/test/kotlin/com/lucas/matches/service/util/LicenseParserTest.kt
package com.lucas.matches.service.util

import com.lucas.matches.service.domain.LicenseType
import com.lucas.matches.service.exception.InvalidLicenseException
import org.junit.jupiter.api.*

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class LicenseParserTest {
    private lateinit var licenseParser: LicenseParser
    @BeforeAll
    fun setup() {
        licenseParser = LicenseParser()
    }

    @Test
    @DisplayName("Parse a valid tournament license")
    fun parseValidTournamentLicense() {
        val licenseString = "TOURNAMENT-t1"
        val licenses = licenseParser.parse(listOf(licenseString))
        Assertions.assertNotNull(licenses)
        Assertions.assertNotNull(licenses[0])
        Assertions.assertEquals(LicenseType.TOURNAMENT, licenses[0].licenseType)
    }

    @Test
    @DisplayName("Parse a valid match license")
    fun parseValidMatchLicense() {
        val licenseString = "MATCH-m1"
        val licenses = licenseParser.parse(listOf(licenseString))
        Assertions.assertNotNull(licenses)
        Assertions.assertNotNull(licenses[0])
        Assertions.assertEquals(LicenseType.MATCH, licenses[0].licenseType)
    }

    @Test
    @DisplayName("Parse multiple valid licenses")
    fun parseMultipleValidLicenses() {
        val licenseStringList = listOf("TOURNAMENT-t1", "MATCH-m1")
        val licenses = licenseParser.parse(licenseStringList)
        Assertions.assertNotNull(licenses)
        Assertions.assertEquals(2, licenses.size)
        Assertions.assertEquals(LicenseType.TOURNAMENT, licenses[0].licenseType)
        Assertions.assertEquals(LicenseType.MATCH, licenses[1].licenseType)
    }

    @Test
    @DisplayName("Parse an invalid license")
    fun parseInvalidLicense() {
        val licenseString = "invalid-m1"
        Assertions.assertThrows(InvalidLicenseException::class.java) { licenseParser.parse(listOf(licenseString)) }
    }

    @Test
    @DisplayName("Parse badly formatted license")
    fun parseBadlyFormattedLicense() {
        val licenseString = "MATCHm1"
        Assertions.assertThrows(InvalidLicenseException::class.java) { licenseParser.parse(listOf(licenseString)) }
    }

    @Test
    @DisplayName("Parse empty license")
    fun parseEmptyLicense() {
        val licenseString = ""
        Assertions.assertThrows(InvalidLicenseException::class.java) { licenseParser.parse(listOf(licenseString)) }
    }
}